### PR TITLE
Initial changes to implement Random short url generation.

### PR DESCRIPTION
**Pull request by raoprasad-zz**

### CODE REVIEW

File: src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java [hunk 1]
```diff
@@ -14,6 +14,8 @@
 import com.firstpenguin.tinyurl.restservice.entity.Url;
 import com.firstpenguin.tinyurl.restservice.repository.RedisUrlRepository;
 import com.firstpenguin.tinyurl.restservice.repository.URLRepository;
+import com.firstpenguin.tinyurl.restservice.services.ShortURLCodeGenerationService;
+import com.firstpenguin.tinyurl.restservice.util.GenerationSpaceExhaustedException;
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java]: import com.firstpenguin.tinyurl.restservice.util.ExhaustedSequenceException ?

File: src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java [hunk 2]
```diff
@@ -25,7 +27,7 @@
     RedisUrlRepository redisUrlRepository;
 
     @Autowired
-    ShortUrlCodeGenerator<String> shortUrlCodeGenerator;
+    ShortURLCodeGenerationService shortUrlGenerationService;
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java]: Good idea to move this to a service

File: src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java [hunk 3]
```diff
@@ -34,19 +36,27 @@ public Url shortUrl(@RequestBody Url url) throws Exception {
         // do we have the URL already stored?
         // TODO: move this call to the Cache
         Url urlFromRepo = urlRepository.findByLongUrlHash(sha1);
+        
+        if (urlFromRepo != null) {
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java]: Good to pull the latest. Code is simplified based on your last review.

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import com.firstpenguin.tinyurl.restservice.util.SequenceGenerationException;
+
+@Service
+public class ShortURLCodeGenerationService {
+	
+	@Autowired
+	RandomSequenceCachingGenerator generator;
+	
+	@Autowired
+    URLRepository urlRepository;
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java]: Good to check number of spaces/tabs. Seeing misaligned code.

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	
+	public String getShortURL() throws SequenceGenerationException {
+		String candidate = "";
+		do
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java]: do {
    candidate = generator.getSequence();
}

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+public class ShortURLCodeGenerationService {
+	
+	@Autowired
+	RandomSequenceCachingGenerator generator;
+	
+	@Autowired
+    URLRepository urlRepository;
+	
+	public String getShortURL() throws SequenceGenerationException {
+		String candidate = "";
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java]: No need to initialize. String can be null?

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+		{
+			candidate = generator.getSequence();
+		}
+		while(isURLInUse(candidate));
+		return candidate;
+	}
+	
+	private boolean isURLInUse(String shortUrl) {
+		try {
+			urlRepository.findById(shortUrl).get();
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java]: Need to check the return value

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/RandomSequenceCachingGenerator.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	private static final int cacheCapacity = 5;
+	private static final int cacheWait = 1000;
+	
+	public RandomSequenceCachingGenerator() {
+		cache = new ArrayBlockingQueue<String>(cacheCapacity);
+	}
+	
+	@Override
+	public void run() {
+		// TODO Auto-generated method stub
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/RandomSequenceCachingGenerator.java:31]: Please remove comment (other places too)

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/URLSequenceGenerator.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+	
+	@Setter
+	String postfix;
+	
+	@Setter
+	@Getter
+	Integer seqLen;
+	
+	@Setter
+	List<Character> choices;
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/URLSequenceGenerator.java]: why do we need this?

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	RandomSequenceCachingGenerator generator;
+	
+	@Autowired
+    URLRepository urlRepository;
+	
+	public String getShortURL() throws SequenceGenerationException {
+		String candidate = "";
+		do
+		{
+			candidate = generator.getSequence();
```
Comment (nnovoice) [inline src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java:33]: can we rename the candidate to candiateSequence?

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/URLSequenceGenerator.java (hunk 9 above)
Comment (nnovoice) [reply src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/URLSequenceGenerator.java]: I get it, this is to store the alphabets. Let us initialize this.

File: src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java (hunk 1 above)
Comment (raoprasad-zz) [reply src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java]: Modified as requested

File: src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java (hunk 3 above)
Comment (raoprasad-zz) [reply src/main/java/com/firstpenguin/tinyurl/restservice/controller/TinyURLController.java]: Pulled and merged.

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java (hunk 5 above)
Comment (raoprasad-zz) [reply src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java]: Not sure what was the comment here?

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java (hunk 6 above)
Comment (raoprasad-zz) [reply src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java]: Sure

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java (hunk 7 above)
Comment (raoprasad-zz) [reply src/main/java/com/firstpenguin/tinyurl/restservice/services/ShortURLCodeGenerationService.java]: This has been refactored in the subsequent commits.

File: src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/RandomSequenceCachingGenerator.java (hunk 8 above)
Comment (raoprasad-zz) [reply src/main/java/com/firstpenguin/tinyurl/restservice/services/generators/RandomSequenceCachingGenerator.java:31]: Done

**Comment by raoprasad-zz** [comment]
Raghav. This is the updated branch for this pull request. Nothing to push from my side. Please merge into master if the changes are appropriate.

**Pull request merged**